tests/js/user-tests.js

```diff
@@ -0,0 +1,279 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2016 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+/* eslint-env es6, node */
+
+'use strict';
+
+const Realm = require('realm');
+const TestCase = require('./asserts');
+const schemas = require('./schemas');
+
+function uuid() {
+  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function (c) {
+    var r = Math.random() * 16 | 0, v = c == 'x' ? r : (r & 0x3 | 0x8);
+    return v.toString(16);
+  });
+}
+
+function assertIsUser(user, isAdmin) {
+  TestCase.assertType(user, 'object');
+  TestCase.assertType(user.token, 'string');
+  TestCase.assertType(user.identity, 'string');
+  TestCase.assertInstanceOf(user, Realm.Sync.User);
+  if (isAdmin !== undefined) {
+    TestCase.assertEqual(user.isAdmin, isAdmin);
+  }
+}
+
+function assertIsSameUser(value, user) {
+  assertIsUser(value);
+  TestCase.assertEqual(value.token, user.token);
+  TestCase.assertEqual(value.identity, user.identity);
+  TestCase.assertEqual(value.isAdmin, user.isAdmin);
+}
+
+function assertIsError(error, code) {
+  TestCase.assertInstanceOf(error, Error, 'The API should return an Error');
+  if (code) {
+    TestCase.assertEqual(error.code, code);
+  }
+}
+
+function assertIsAuthError(error, code, type) {
+  TestCase.assertInstanceOf(error, Realm.Sync.AuthError, 'The API should return an AuthError');
+  if (code) {
+    TestCase.assertEqual(error.code, code);
+  }
+  if (type) {
+    TestCase.assertEqual(error.type, type);
+  }
+}
+
+function rejectOnError(error, reject) {
+  if (error) {
+    reject(error);
+  }
+}
+
+function failOnError(error) {
+  if (error) {
+    throw new Error(`Error ${error} was not expected`);
+  }
+}
+
+// Test the given requestFunc, passing it the given callback after it's been wrapped appropriately.
+// This function makes sure that errors thrown in the async callback rejects the promise (making tests actually run).
+function callbackTest(requestFunc, callback) {
+  return new Promise((resolve, reject) => {
+    function callbackWrapper() {
+      try {
```

Comment:
I don't think so. In my opinion this construct unnecessarily obfuscates the program flow. That was also the reason why I noticed it at all in the first place.